File: 2024/src/day06.rs
use crate::{read_input, Grid, Vec2};
use std::collections::HashSet;

fn find_start(grid: &Grid<char>) -> Option<Vec2> {
    if let Some((x, y)) = grid.find(&'^') {
        return Some(Vec2::new(x, y));
    }
    None
}

fn sum_visited_fields(input: String) -> usize {
    let grid = Grid::from_string(&input);

    if let Some(mut current_pos) = find_start(&grid) {
        let mut visited = HashSet::new();
        let mut direction = Vec2::new(0, -1);
        let mut next_pos = current_pos.add(&direction);
        visited.insert(current_pos);

        while let Some(next_char) = grid.get(next_pos.x, next_pos.y) {
            // Update direction based on current character
            match next_char {
                '#' => {
                    // Next char is obstacle, rotate right 90°
                    direction = Vec2::new(-direction.y, direction.x);
                }
                _ => {
                    // Continue in same direction
                    current_pos = next_pos;
                    visited.insert(current_pos);
                }
            }

            next_pos = current_pos.add(&direction);
        }
        return visited.len();
    }
    0
}

pub fn part1() -> usize {
    sum_visited_fields(read_input(6))
}

pub fn part2() -> i64 {
    0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example_part1() {
        let input = String::from(
            r#"....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#..."#,
        );
        assert_eq!(sum_visited_fields(input), 41 as usize);
    }
}

File: 2024/go/days/day06/day06.go
package day06

import "aoc/2024/go/lib"

func findStart(grid *lib.Grid) lib.Vec2 {
	for y := 0; y < grid.Height(); y++ {
		for x := 0; x < grid.Width(); x++ {
			if grid.Get(x, y) == '^' {
				return lib.Vec2{X: x, Y: y}
			}
		}
	}
	panic("start position not found")
}

func trackGuardFields(grid *lib.Grid, start lib.Vec2, obstacle *lib.Vec2) (visited map[lib.Vec2]bool, trapped bool) {
	current := start
	visited = make(map[lib.Vec2]bool)

	visitedByDirection := make(map[lib.Vec2]map[lib.Vec2]bool)
	directions := []lib.Vec2{
		{X: 0, Y: -1},
		{X: 1, Y: 0},
		{X: 0, Y: 1},
		{X: -1, Y: 0},
	}
	for _, direction := range directions {
		visitedByDirection[direction] = make(map[lib.Vec2]bool)
	}

	direction := lib.Vec2{X: 0, Y: -1}
	for {
		next := current.Add(direction)
		visited[current] = true
		if visitedByDirection[direction][current] {
			return visited, true
		}
		visitedByDirection[direction][current] = true
		if !grid.Contains(next.X, next.Y) {
			break
		}
		if grid.Get(next.X, next.Y) == '#' || (obstacle != nil && *obstacle == next) {
			// Next char is obstacle, rotate right 90°
			direction = lib.Vec2{X: -direction.Y, Y: direction.X}
		} else {
			// Continue in this direction
			current = next
		}
	}
	return visited, false
}

func Part1() interface{} {
	input, _ := lib.ReadInput(6)
	grid := lib.NewGrid(input)
	start := findStart(grid)
	fieldsVisited, _ := trackGuardFields(grid, start, nil)
	return len(fieldsVisited)
}

func countLoopCausingObstacles(grid *lib.Grid, start lib.Vec2, visited map[lib.Vec2]bool) int {
	loopCausingObstacles := 0
	for position := range visited {
		if position == start {
			continue
		}
		_, isLoop := trackGuardFields(grid, start, &position)
		if isLoop {
			loopCausingObstacles++
		}
	}
	return loopCausingObstacles
}

func Part2() interface{} {
	input, _ := lib.ReadInput(6)
	grid := lib.NewGrid(input)
	start := findStart(grid)
	fieldsVisited, _ := trackGuardFields(grid, start, nil)
	return countLoopCausingObstacles(grid, start, fieldsVisited)
}
